File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        brackets = {
            "(": ")",
            "[": "]",
            "{": "}"
        }
        stack = []

        for bracket in s:
            if stack and stack[-1] in brackets and brackets[stack[-1]] == bracket:
                stack.pop()
            else:
                stack.append(bracket)
            
        return not stack

File: 20-valid-parentheses/valid-parentheses.ts
function isValid(s: string): boolean {
    const validMap = {
        ")": "(",
        "}": "{",
        "]": "["
    }    
    const openParentheses = ["(", "{", "["]
    const stack = [];

    for (const char of s) {
        if (openParentheses.includes(char)) {
            stack.push(char)
        } else {
            if (stack.at(-1) === validMap[char]) {
                stack.pop()
            } else {
                return false
            }
        }
    }

    return stack.length === 0
};